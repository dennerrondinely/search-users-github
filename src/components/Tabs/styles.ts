import styled, { css } from 'styled-components';

type HeaderItemProps = {
  active: boolean;
};

export const Wrapper = styled.div`
  ${({ theme }) => css`
    width: 100%;
    max-width: ${theme.grid.container};
    margin-left: auto;
    margin-right: auto;
    padding: 0;
  `}
`;

export const Header = styled.ul`
  list-style: none;
  display: flex;
`;

export const HeaderItem = styled.li<HeaderItemProps>`
  position: relative;
  cursor: pointer;
  transform: translateY(0.1rem);
  ${({ theme, active }) => css`
    padding: ${theme.spacings.xxsmall} ${theme.spacings.small};
    color: ${active ? theme.colors.secondary : theme.colors['gray-bright']};
    gap: ${theme.spacings.small};

    &::before {
      content: '';
      height: 0.6rem;
      width: 0.6rem;
      border-radius: 0.3rem;
      position: absolute;
      top: 50%;
      transform: translateY(-50%);
      left: ${theme.spacings.minimal};
      transition: background-color 0.3s;
      background-color: ${active
        ? theme.colors.secondary
        : theme.colors.transparent};
    }
  `}
`;

export const TabPane = styled.div`
  ${({ theme }) => css`
    padding: ${theme.spacings.medium} 0;
  `}
`;
