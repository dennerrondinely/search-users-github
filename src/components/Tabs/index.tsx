import React, { ReactNode, useEffect, useState } from 'react';

import * as S from './styles';

export type TabsProps = {
  activeTab?: string;
  children: ReactNode;
};

const Tabs = ({ children, activeTab }: TabsProps) => {
  const [tabHeader, setTabHeader] = useState<string[]>([]);
  const [childContent, setChildConent] = useState({});
  const [active, setActive] = useState(activeTab);

  useEffect(() => {
    const headers: string[] = [];
    const childCnt = {};
    React.Children.forEach(children, (element) => {
      if (!React.isValidElement(element)) return;
      const { name } = element.props;
      headers.push(name);
      childCnt[name] = element;
    });

    setTabHeader(headers);
    setChildConent({ ...childCnt });
  }, [children]);

  const changeTab = (name: string) => {
    setActive(name);
  };
  return (
    <S.Wrapper>
      <S.Header>
        {tabHeader.map((item) => (
          <S.HeaderItem
            active={item === active}
            key={item}
            onClick={() => changeTab(item)}
          >
            {/* <Typography
              variant="body-2"
              weight="bold"
              color={item === active ? 'secondary' : 'gray-1'}
            >
              {item}
            </Typography> */}
            <span>{item}</span>
          </S.HeaderItem>
        ))}
      </S.Header>
      {Object.keys(childContent).map(
        (key) => key === active && childContent[key]
      )}
    </S.Wrapper>
  );
};

export type TabPaneProps = {
  name: string;
  children: ReactNode;
};

export const TabPane = ({ children }: TabPaneProps) => (
  <S.TabPane>{children}</S.TabPane>
);

export default Tabs;
