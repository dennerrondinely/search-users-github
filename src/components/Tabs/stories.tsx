import React from 'react';
import { Story, Meta } from '@storybook/react/types-6-0';

import Tabs, { TabsProps, TabPane } from './index';

export default {
  title: 'Tabs',
  component: Tabs
} as Meta;

export const Default: Story<TabsProps> = (args) => (
  <Tabs {...args}>
    <TabPane name="Repositorios" key="1">
      Repositorios
    </TabPane>
    <TabPane name="Repositorios com Estrelou" key="2">
      Repositorios com Estrelou
    </TabPane>
  </Tabs>
);

Default.args = {
  activeTab: 'Repositorios'
};
