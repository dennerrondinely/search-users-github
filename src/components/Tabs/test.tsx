import React from 'react';
import { fireEvent, screen } from '@testing-library/react';

import { render } from 'utils/tests/helpers';

import Tabs, { TabPane } from './index';

describe('<Tabs />', () => {
  const Component = () => (
    <Tabs>
      <TabPane name="Informação">Content of Tab Pane 1</TabPane>
      <TabPane name="Historico do Cliente">Content of Tab Pane 2</TabPane>
      <TabPane name="contrato social">Content of Tab Pane 3</TabPane>
    </Tabs>
  );

  it('should render the Tabs', () => {
    const { container } = render(<Component />);

    expect(container).toMatchSnapshot();
  });

  it('should change Tabs', () => {
    const { container } = render(<Component />);

    fireEvent.click(screen.getByText('Historico do Cliente'));
    expect(container).toHaveTextContent('Content of Tab Pane 2');

    fireEvent.click(screen.getByText('contrato social'));
    expect(container).toHaveTextContent('Content of Tab Pane 3');

    fireEvent.click(screen.getByText('Informação'));
    expect(container).toHaveTextContent('Content of Tab Pane 1');
  });

  it('should valid is a element', () => {
    const { container } = render(<Tabs>teste</Tabs>);

    expect(container).not.toContainHTML('li');
  });
});
