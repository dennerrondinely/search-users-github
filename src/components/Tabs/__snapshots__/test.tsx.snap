// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tabs /> should render the Tabs 1`] = `
.c0 {
  width: 100%;
  max-width: 114.4rem;
  margin-left: auto;
  margin-right: auto;
  padding: 0;
}

.c1 {
  list-style: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  max-width: 60rem;
  margin: 0 auto;
  padding: 0 2.4rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c2 {
  position: relative;
  cursor: pointer;
  -webkit-transform: translateY(0.1rem);
  -ms-transform: translateY(0.1rem);
  transform: translateY(0.1rem);
  padding: 0.8rem 2.4rem;
  color: #A9A5A2;
  gap: 2.4rem;
}

.c2::before {
  content: '';
  height: 0.6rem;
  width: 0.6rem;
  border-radius: 0.3rem;
  position: absolute;
  top: 50%;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
  left: 0.4rem;
  -webkit-transition: background-color 0.3s;
  transition: background-color 0.3s;
  background-color: #0000;
}

<div>
  <div
    class="c0"
  >
    <ul
      class="c1"
    >
      <li
        class="c2"
      >
        <span>
          Informação
        </span>
      </li>
      <li
        class="c2"
      >
        <span>
          Historico do Cliente
        </span>
      </li>
      <li
        class="c2"
      >
        <span>
          contrato social
        </span>
      </li>
    </ul>
  </div>
</div>
`;
